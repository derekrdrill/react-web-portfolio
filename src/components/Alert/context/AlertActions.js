export const handleAlert = (msg, title, type, alertDispatch) => {
  alertDispatch({
    type: 'SET_ALERT',
    fadeOut: false,
    payload: { msg: msg, title: title, type: type },
  });

  setTimeout(() => alertDispatch({ type: 'FADE_ALERT', fadeOut: true }), 2000);
  setTimeout(() => alertDispatch({ type: 'REMOVE_ALERT' }), 2400);
};
