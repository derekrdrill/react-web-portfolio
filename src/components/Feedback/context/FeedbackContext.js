import React, { createContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

export const FeedbackContext = createContext();

export const FeedbackProvider = ({ children }) => {
  const [feedbackItems, setFeedbackItems] = useState([]);
  const [feedbackEdit, setFeedbackEdit] = useState({
    item: {},
    edit: false,
  });

  const getFeedbackItems = async () => {
    const response = await fetch('feedbackItems').catch(e => console.warn(e));

    if (response.ok) {
      const feedbackData = await response.json();
      setFeedbackItems(feedbackData);
    } else {
      console.warn(Promise.reject(response));
    }
  };

  const addFeedbackItem = async newFeedbackItem => {
    const response = await fetch('addFeedbackItem', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(newFeedbackItem),
    }).catch(e => console.warn(e));

    if (response.ok) {
      const feedbackData = await response.json();
      setFeedbackItems([...feedbackItems, feedbackData]);
    } else {
      console.warn(Promise.reject(response));
    }
  };

  const deleteFeedbackItem = async deleteID => {
    const response = await fetch('deleteFeedbackItem', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ id: deleteID }),
    }).catch(e => console.warn(e));

    if (response.ok) {
      const feedbackData = await response.json();
      const feedbackDataId = feedbackData._id;
      setFeedbackItems(feedbackItems.filter(item => item._id !== feedbackDataId));
    } else {
      console.warn(Promise.reject(response));
    }
  };

  const setFeedbackEditItems = (item, edit) => setFeedbackEdit({ item, edit: edit });

  const updateFeedbackItem = async updateItem => {
    const response = await fetch('updateFeedbackItem', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(updateItem),
    });

    if (response.ok) {
      const feedbackData = await response.json();
      const feedbackDataId = feedbackData._id;

      setFeedbackItems(
        feedbackItems.map(feedbackItem =>
          feedbackItem._id === feedbackDataId
            ? Object.assign({}, feedbackItem, {
                rating: feedbackData.rating,
                text: feedbackData.text,
              })
            : feedbackItem,
        ),
      );

      setFeedbackEdit({ item: {}, edit: false });
    } else {
      console.warn(Promise.reject(response));
    }
  };

  useEffect(
    /* istanbul ignore next */
    () => getFeedbackItems(),
    [],
  );

  return (
    <FeedbackContext.Provider
      value={{
        feedbackEdit,
        feedbackItems,
        addFeedbackItem,
        deleteFeedbackItem,
        setFeedbackEditItems,
        updateFeedbackItem,
      }}
    >
      {children}
    </FeedbackContext.Provider>
  );
};

FeedbackProvider.propTypes = {
  children: PropTypes.node,
};
