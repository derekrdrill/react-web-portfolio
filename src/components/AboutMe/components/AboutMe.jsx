import React from 'react';
import styled, { createGlobalStyle } from 'styled-components';
import { Container, Row, Col } from 'react-bootstrap';
import { Typography, List, ListItem } from '@mui/material';

import { DarkLightModeContext } from '../../DarkLightMode/context/DarkLightModeContext';

import { ABOUT_ME_TEXT } from '../constants/ABOUT_ME_TEXT';
import bitmojiChillinWithBirds from '../../../assets/bitmoji_chillinWithBirds1.png';

export const AboutMe = ({ id }) => {
  // test change

  const { darkMode } = React.useContext(DarkLightModeContext);

  return (
    <>
      <PageBodyStyle darkMode={darkMode} />
      <AboutMeContainer darkMode={darkMode} id={id} fluid>
        <Row>
          <Col className='py-3'>
            <AboutMeTitle className='text-center' component='h1' darkMode={darkMode} variant='h3'>
              ABOUT ME
            </AboutMeTitle>
          </Col>
        </Row>
        <Row>
          <AboutMeTextContainer
            darkMode={darkMode}
            xs={{ span: 10, order: 'last', offset: 1 }}
            lg={{ span: 5, order: 'first', offset: 1 }}
          >
            <AboutMeText darkMode={darkMode} variant='subtitle1' component='p'>
              {ABOUT_ME_TEXT[1].text}
            </AboutMeText>
            <AboutMeText darkMode={darkMode} variant='subtitle1' component='p'>
              {ABOUT_ME_TEXT[2].text}
            </AboutMeText>
            <List style={{ color: 'white' }}>
              <AboutMeListItem darkMode={darkMode}>{ABOUT_ME_TEXT[3].text}</AboutMeListItem>
              <AboutMeListItem darkMode={darkMode}>{ABOUT_ME_TEXT[4].text}</AboutMeListItem>
              <AboutMeListItem darkMode={darkMode}>{ABOUT_ME_TEXT[5].text}</AboutMeListItem>
              <AboutMeListItem darkMode={darkMode}>{ABOUT_ME_TEXT[6].text}</AboutMeListItem>
            </List>
            <AboutMeText darkMode={darkMode} variant='subtitle1' component='p'>
              {ABOUT_ME_TEXT[7].text}
            </AboutMeText>
            <AboutMeText darkMode={darkMode} variant='subtitle1' component='p'>
              {ABOUT_ME_TEXT[8].text}
            </AboutMeText>
            <AboutMeText darkMode={darkMode} variant='subtitle1' component='p'>
              {ABOUT_ME_TEXT[9].text}
            </AboutMeText>
          </AboutMeTextContainer>
          <AboutMeBitmojiContainer xs={12} lg={6}>
            <img src={bitmojiChillinWithBirds} alt='Chillin with birds' />
          </AboutMeBitmojiContainer>
        </Row>
      </AboutMeContainer>
    </>
  );
};

const PageBodyStyle = createGlobalStyle(({ darkMode }) => ({
  body: {
    backgroundColor: darkMode ? '#030200' : 'whitesmoke',
  },
}));

export const AboutMeContainer = styled(Container)(({ darkMode }) => ({
  padding: '30px 50px 200px 50px',
  borderTop: '2px solid transparent',
  borderImage: darkMode
    ? 'linear-gradient(to right, rgba(248, 184, 255, 1), skyblue, gainsboro)'
    : 'linear-gradient(to right, rgba(176, 52, 197, 1), rgba(86, 206, 210, 1))',
  borderImageSlice: 1,
}));

export const AboutMeTitle = styled(Typography)(({ darkMode }) => ({
  fontFamily: 'Shizuru',
  fontWeight: darkMode ? 'normal' : 'bold',
  color: darkMode ? 'gainsboro' : 'darkslategray',
}));

export const AboutMeTextContainer = styled(Col)(({ darkMode }) => ({
  padding: 40,
  height: 550,
  overflowY: 'auto',
  border: '1px dashed #383838',
  borderRadius: 2,
  backgroundColor: darkMode ? '#303030' : '#eee5c1',
}));

export const AboutMeBitmojiContainer = styled(Col)({
  textAlign: 'center',
  paddingTop: 75,
  paddingBottom: 20,
});

export const AboutMeText = styled(Typography)(({ darkMode }) => ({
  fontFamily: 'Kufam',
  color: darkMode ? 'gainsboro' : '#333333',
  marginBottom: 15,
}));

export const AboutMeListItem = styled(ListItem)(({ darkMode }) => ({
  fontFamily: 'Kufam',
  color: darkMode ? 'gainsboro' : '#333333',
}));
