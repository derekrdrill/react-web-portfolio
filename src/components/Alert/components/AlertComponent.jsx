import React from 'react';
import styled, { css, keyframes } from 'styled-components';
import { Alert, AlertTitle } from '@mui/material';
import { AlertContext } from '../context/AlertContext';
import { DarkLightModeContext } from '../../DarkLightMode/context/DarkLightModeContext';

export const AlertComponent = () => {
  const { alert, fadeOut } = React.useContext(AlertContext);
  const { darkMode } = React.useContext(DarkLightModeContext);

  return (
    alert && (
      <StyledAlert darkMode={darkMode} severity={alert.type} $fadeOut={fadeOut}>
        <AlertTitle>{alert.title}</AlertTitle>
        {alert.msg}
      </StyledAlert>
    )
  );
};

const fadeOut = keyframes({
  '0%': {
    opacity: 1,
  },
  '100%': {
    opacity: 0,
  },
});

export const StyledAlert = styled(Alert)`
  animation-name: ${props =>
    props.$fadeOut
      ? css`
          ${fadeOut}
        `
      : 'none'};
  animation-duration: 500ms;
  animation-direction: linear;
`;
