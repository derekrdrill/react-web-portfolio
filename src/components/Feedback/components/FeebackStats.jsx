import React from 'react';
import styled from 'styled-components';
import { FeedbackContext } from '../context/FeedbackContext';

export const getAverage = feedbackItems =>
  feedbackItems.reduce((acc, current) => {
    return acc + current.rating;
  }, 0) / feedbackItems.length;

export const getAverageFixed = feedbackItems =>
  isNaN(getAverage(feedbackItems))
    ? 0
    : getAverage(feedbackItems) % 1 > 0
    ? getAverage(feedbackItems).toFixed(1)
    : getAverage(feedbackItems);

export const FeedbackStats = () => {
  const { feedbackItems } = React.useContext(FeedbackContext);

  return (
    <FeedbackStatsDiv>
      <h4>{feedbackItems.length} reviews</h4>
      <h4>Average rating: {getAverageFixed(feedbackItems)}</h4>
    </FeedbackStatsDiv>
  );
};

const FeedbackStatsDiv = styled.div({
  alignItems: 'center',
  color: '#fff',
  display: 'flex',
  justifyContent: 'space-between',
  margin: '20px 30px',
});
