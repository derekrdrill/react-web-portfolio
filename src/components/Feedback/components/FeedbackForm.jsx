import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Button, FormControl, FormControlLabel, Radio, RadioGroup, TextField, Typography } from '@mui/material';
import { Card } from '../../Card/Card';
import { FeedbackContext } from '../context/FeedbackContext';

export const FeedbackForm = ({ maxRank }) => {
  const menuItems = new Array(maxRank + 1).fill(null);
  const { addFeedbackItem, feedbackEdit, updateFeedbackItem } = useContext(FeedbackContext);

  const [rating, setRating] = useState(null);
  const [review, setReview] = useState('');

  const handleSetRating = (e, value) => setRating(Number(value));
  const handleSetReview = e => setReview(e.currentTarget.value);
  const handleSubmitReview = () => {
    if (!feedbackEdit.edit) {
      addFeedbackItem({
        rating: rating,
        text: review,
      });
    } else {
      updateFeedbackItem({
        _id: feedbackEdit.item.id,
        rating: rating,
        text: review,
      });
    }

    setRating(null);
    setReview('');
  };

  useEffect(() => {
    if (feedbackEdit.edit) {
      setReview(feedbackEdit.item.text);
      setRating(feedbackEdit.item.rating);
    }
  }, [feedbackEdit]);

  return (
    <Card>
      <h2>Feedback Form</h2>
      <Typography component='p' variant='subtitle1'>
        Please rate your experience
      </Typography>
      <FormControl fullWidth>
        <RadioGroup onChange={handleSetRating}>
          <RadioButtonsContainer>
            {menuItems.map((menuItem, menuItemKey) => (
              <div key={menuItemKey}>
                <RatingRadioControl
                  checked={menuItemKey === rating}
                  control={<Radio />}
                  label={menuItemKey}
                  value={menuItemKey}
                />
              </div>
            ))}
          </RadioButtonsContainer>
        </RadioGroup>
      </FormControl>
      <TextField
        fullWidth
        label={feedbackEdit.edit ? 'Edit review' : 'Write a review'}
        minRows={2}
        multiline
        onChange={handleSetReview}
        placeholder={feedbackEdit.edit ? 'Edit review' : 'Write a review'}
        value={review}
      />
      <ButtonContainer>
        <Button
          disabled={!review || typeof rating !== 'number'}
          onClick={handleSubmitReview}
          size='small'
          variant='outlined'
        >
          {feedbackEdit.edit ? 'Update ' : 'Submit '}
          Review
        </Button>
      </ButtonContainer>
    </Card>
  );
};

FeedbackForm.propTypes = {
  maxRank: PropTypes.number.isRequired,
};

FeedbackForm.defaultProps = {
  maxRank: 10,
};

const ButtonContainer = styled.div({
  display: 'flex',
  justifyContent: 'flex-end',
  padding: '15px 0',
});

const RadioButtonsContainer = styled.div({
  display: 'flex',
  flexWrap: 'wrap',
  justifyContent: 'center',
  margin: '40px 30px',
});

const RatingRadioControl = styled(FormControlLabel)(({ checked }) => [
  {
    '.MuiRadio-root': {
      display: 'none',
    },
    ':hover': {
      backgroundColor: '#FF77BC',
      color: 'beige',
    },
    backgroundColor: 'lavender',
    borderRadius: 50,
    color: '#333',
    minWidth: 12,
    padding: '25px 30px',
    cursor: 'pointer',
  },
  checked && {
    ':hover': {
      backgroundColor: '#be7dff',
    },
    backgroundColor: '#be7dff',
    color: 'beige',
  },
]);
