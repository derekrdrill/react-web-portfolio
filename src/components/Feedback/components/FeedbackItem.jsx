import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Button, Typography } from '@mui/material';
import { Card } from '../../Card/Card';
import { BasicModal } from '../../Modals/BasicModal';
import { FeedbackContext } from '../context/FeedbackContext';
import { FaTimes, FaEdit } from 'react-icons/fa';

export const FeedbackItem = ({ bgColorNumDisplay, id, rating, text }) => {
  const { deleteFeedbackItem, setFeedbackEditItems } = React.useContext(FeedbackContext);
  const [modalOpen, setModalOpen] = useState(false);
  const handleModalOpen = () => setModalOpen(true);
  const handleModalClose = () => setModalOpen(false);

  return (
    <Card>
      <NumberDisplay bgColor={bgColorNumDisplay}>
        <Typography variant='subtitle' component='h6'>
          {rating}
        </Typography>
      </NumberDisplay>
      <ActionButtonsDisplay>
        <ActionsButtonContainer>
          <Button
            color='secondary'
            fullWidth
            onClick={() =>
              setFeedbackEditItems(
                {
                  id: id,
                  text: text,
                  rating: rating,
                },
                true,
              )
            }
          >
            <FaEdit color='purple' />
          </Button>
        </ActionsButtonContainer>
        <ActionsButtonContainer>
          <Button color='secondary' fullWidth onClick={handleModalOpen}>
            <FaTimes color='purple' />
          </Button>
        </ActionsButtonContainer>
      </ActionButtonsDisplay>
      <TextDisplay>
        <Typography variant='subtitle1' component='p'>
          {text}
        </Typography>
      </TextDisplay>
      <BasicModal
        cancelButtonText='I changed my mind'
        handleClose={handleModalClose}
        handleSubmit={() => deleteFeedbackItem(id)}
        open={modalOpen}
        submitButtonText='Yes, delete it'
      >
        <Typography paragraph>Are you sure you want to delete this feedback record?</Typography>
      </BasicModal>
    </Card>
  );
};

FeedbackItem.propTypes = {
  bgColorNumDisplay: PropTypes.string,
  id: PropTypes.string,
  rating: PropTypes.number,
  text: PropTypes.string,
};

export const NumberDisplay = styled.div(({ bgColor }) => ({
  position: 'relative',
  background: '#f4f4f4',
  width: 50,
  height: 50,
  padding: 4,
  textAlign: 'center',
  borderRadius: '50%',
  fontSize: 30,
  border: '1px #eee solid',
  transition: '0.3s',
  color: '#FFF',
  backgroundColor: bgColor,
  transform: 'translate(-60px, -50px)',
  '.MuiTypography-root': {
    paddingTop: 12,
  },
}));

const ActionsButtonContainer = styled.div({
  display: 'flex',
  justifyContent: 'flex-end',
  maxWidth: 50,
});

const ActionButtonsDisplay = styled.div({
  position: 'absolute',
  top: 10,
  right: 1,
  cursor: 'pointer',
  background: 'none',
  border: 'none',
  display: 'flex',
  justifyContent: 'flex-end',
});

const TextDisplay = styled.div({
  transform: 'translate(0px, -40px)',
});
