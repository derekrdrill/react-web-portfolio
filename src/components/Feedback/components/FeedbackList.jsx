import React from 'react';
import { FeedbackContext } from '../context/FeedbackContext';
import { FeedbackItem } from './FeedbackItem';

export const FeedbackList = () => {
  const { feedbackItems } = React.useContext(FeedbackContext);

  return (
    <>
      {feedbackItems &&
        feedbackItems.map(feedbackItem => (
          <FeedbackItem
            bgColorNumDisplay='#FF77BC'
            id={feedbackItem._id}
            key={feedbackItem._id}
            rating={feedbackItem.rating}
            text={feedbackItem.text}
          />
        ))}
    </>
  );
};
