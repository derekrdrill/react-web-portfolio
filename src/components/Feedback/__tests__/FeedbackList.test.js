import 'regenerator-runtime/runtime';

import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { FeedbackList } from '../components/FeedbackList';
import { FeedbackProvider } from '../context/FeedbackContext';

const renderer = new ShallowRenderer();

describe('Feedback List tests', () => {
  it('renders correctly', () => {
    renderer.render(
      <FeedbackProvider>
        <FeedbackList />
      </FeedbackProvider>,
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
