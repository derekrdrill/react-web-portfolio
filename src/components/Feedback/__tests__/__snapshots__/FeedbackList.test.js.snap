// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback List tests renders correctly 1`] = `
<React.Fragment>
  <FeedbackItem
    bgColorNumDisplay="#FF77BC"
    id={1}
    rating={10}
    text="Nice"
  />
  <FeedbackItem
    bgColorNumDisplay="#FF77BC"
    id={2}
    rating={9}
    text="Pretty Good"
  />
</React.Fragment>
`;
